source/com.microsoft.tfs.client.common.ui/src/com/microsoft/tfs/client/common/ui/protocolhandler/ProtocolHandler.java

```diff
@@ -0,0 +1,331 @@
+    public final static String PROTOCOL_HANDLER_SCHEME = "vsoi"; //$NON-NLS-1$
+
+    private static final ProtocolHandler instance = new ProtocolHandler();
+
+    private URI protocolHandlerUri;
+
+    private boolean isParsed = false;
+    private boolean isAvailable = false;
+    private String cloneUrl;
+    private String encoding;
+    private String branchName;
+    private String serverUrl;
+    private String repository;
+    private String project;
+
+    public ProtocolHandler() {
+        protocolHandlerUri = findProtocolHandlerUriArgument(org.eclipse.core.runtime.Platform.getApplicationArgs());
+    }
+
+    public static ProtocolHandler getInstance() {
+        return instance;
+    }
+
+    public boolean hasProtocolHandlerRequest() {
+        return isParsed ? isAvailable : protocolHandlerUri != null;
+    }
+
+    public String getProtocolHandlerServer() {
+        tryParseProtocolHandlerUri();
+        return isAvailable ? serverUrl : StringUtil.EMPTY;
+    }
+
+    public String getProtocolHandlerProject() {
+        tryParseProtocolHandlerUri();
+        return isAvailable ? project : StringUtil.EMPTY;
+    }
+
+    public String getProtocolHandlerBranch() {
+        tryParseProtocolHandlerUri();
+        return isAvailable ? branchName : StringUtil.EMPTY;
+    }
+
+    public String getProtocolHandlerRepository() {
+        tryParseProtocolHandlerUri();
+        return isAvailable ? repository : StringUtil.EMPTY;
+    }
+
+    public String getProtocolHandlerCloneUrl() {
+        tryParseProtocolHandlerUri();
+        return isAvailable ? cloneUrl.toString() : StringUtil.EMPTY;
+    }
+
+    public String getProtocolHandlerEncodedCloneUrl() {
+        tryParseProtocolHandlerUri();
+        return isAvailable ? URIUtils.newURI(cloneUrl).toASCIIString() : StringUtil.EMPTY;
+    }
+
+    private URI findProtocolHandlerUriArgument(final String[] applicationArgs) {
+        if (applicationArgs == null) {
+            return null;
+        }
+        boolean found = false;
+
+        /*
+         * @formatter:off
+         * We're looking for a protocol handler argument among all command line
+         * arguments passed by eclipse.launcher to the Eclipse application. The
+         * protocol handler argument should have the following syntax:
+         * 
+         * -clonefromtfs <uri>
+         * 
+         * At this point we do not parse the value of the argument. We'll do it later.
+         * @formatter:on
+         */
+
+        for (final String arg : applicationArgs) {
+            if (found) {
+                log.info(MessageFormat.format(
+                    "Found the protocol handler argument: {0} {1}", //$NON-NLS-1$
+                    PROTOCOL_HANDLER_ARG,
+                    arg));
+
+                try {
+                    return URIUtils.newURI(arg);
+                } catch (final Exception e) {
+                    log.error("   Incorrect URL in the protocol handler argument", e); //$NON-NLS-1$
```

Comment:
Yes, it is. By idea, in the log it will look like
Found the protocol handler argument ....
       Incorrect URL in the protocol handler argument
It's nice, isn't it? :-)